design/004-datafrom-key-rewrite.md

```diff
@@ -0,0 +1,103 @@
+```yaml
+---
+title: dataFrom key rewrite
+version: v1alpha1
+authors: 
+creation-date: 2022-05-25
+status: draft
+---
+```
+
+# Datafrom Key Rewrite
+
+## Table of Contents
+
+<!-- toc -->
+// autogen please
+<!-- /toc -->
+
+
+## Summary
+DataFrom key rewrite aims to give users the possibility to rewrite secret keys before manipulating into a template. This feature is specially useful with the addition of `dataFrom.find` introduced in v0.5.0, as a way to allow users to rewrite secret keys in batches, without knowing the full content of the key.
+
+## Motivation
+It is hard to administer key creation without having weird names that convolutes with the path. It is natural to have a feature to allow secret Keys manipulation both from the lifecycle of the secret management and as part of the templating feature.
+
+### Goals
+- CRD Design for dataFrom key rewrite
+
+### Non-Goals
+Do not implement secret values rewrite, nor handle with conversion.
+
+### Terminology
+- Secret Key: the kubernetes secret key, aka the `string` in `map[string][]byte` in the codebase used to render Kubernetes Secrets.
+
+## Proposal
+
+Add a rewrite step after applying either `GetSecretMap` or `GetAllSecrets`, but before applying any conversion logic.
+
+### User Stories
+1. As an user I want to be able to remove paths from my Secret Keys
+2. As an user I want to be able to customize secret keys to have more meaningful value to the tenants consuming it.
+3. As an user I want external-secrets to tell me if there is any rewrite conflict and not to overwrite secrets information.
+
+### API
+Proposed CRD changes:
+
+```yaml
+apiVersion: external-secrets.io/v1beta1
+kind: ExternalSecret
+metadata:
+  name: sample
+  namespace: default
+spec:
+  refreshInterval: 1m
+  target:
+   name: foobar
+  secretStoreRef:
+    name: secretstore-sample
+    kind: SecretStore
+  dataFrom:
+    - extract:
+        key: /foo/bar
+      rewrite:
```

Comment:
I might have misunderstood your comment :sweat_smile:  AFAIK the type that they share in common is the `extract` content as a `Extract: dataRemoteRef{}`. The `dataFrom` itself is a `DataFrom: []dataFromRemoteRef` so it wouldn't be injected to `data[]`. The idea behind this would be to add the rewrite type in there (as in parallel with `extract` and `find`) and apply this logic to the controller level, so we do not have to send this information to any provider call.

Reply:
Oh, right! misread the yaml indent :sweat_smile: